src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -82,6 +82,46 @@ public partial class Activity : IDisposable
         private LinkedList<ActivityEvent>? _events;
         private Dictionary<string, object>? _customProperties;
         private string? _displayName;
+        private ActivityStatusCode _statusCode;
+        private string? _statusDescription;
+
+        /// <summary>
+        /// Gets status code of the current activity object.
+        /// </summary>
+        public ActivityStatusCode Status => _statusCode;
+
+        /// <summary>
+        /// Gets the status descrition of the current activity object.
+        /// </summary>
+        public string? StatusDescription => _statusDescription;
+
+        /// <summary>
+        /// Sets the status code and description on the current activity object.
+        /// </summary>
+        /// <param name="code">The status code</param>
+        /// <param name="description">The error status descrition</param>
+        /// <returns>'this' for convenient chaining</returns>
+        /// <remarks>
+        /// It is not allowed to set the status code to Unset value.
+        /// The description parameter is allowed to be non-null value only when passing 'Error' status code.
+        /// </remarks>
+        public Activity SetStatus(ActivityStatusCode code, string? description = null)
+        {
+            if (code <= ActivityStatusCode.Unset || code > ActivityStatusCode.Error)
+            {
+                throw new ArgumentException(SR.Format(SR.InvalidStatusCode, code), nameof(code));
```

Comment:
FWIW, I'm still not clear on why it's deemed acceptable to transition back and forth between Ok and Error but why it's not ok to transition to Unset.

Reply:
@cijothomas @reyang may answer this better.

The [specs](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md) says: `Description MUST be IGNORED for StatusCode Ok & Unset values.`. This means it is allowed to set `Unset`. Also, my current implementation is throwing in such cases. Should we just ignore the description in such cases as the specs suggest?

Reply:
If the spec says these conditions are ok,  and we're adding these to support the spec, I don't know why we'd add our own artificial limitations. (At that point,  you don't even need this method and could just put setters on the properties.)

Reply:
I think we still need SetStatus. The reason is we need to reset the description message to `null` when setting the code to something different than `Error`. That is what the specs is requiring.